oximeter/db/src/client.rs

```diff
@@ -267,14 +271,301 @@ impl Client {
+                context: format!(
+                    "Failed to read schema directory '{}'",
+                    dir.display()
+                ),
+                err,
+            })?;
+        let mut versions = BTreeSet::new();
+        debug!(log, "reading entries from schema dir"; "dir" => dir.display());
+        while let Some(entry) =
+            rd.next_entry().await.map_err(|err| Error::ReadSchemaDir {
+                context: String::from("Failed to read directory entry"),
+                err,
+            })?
+        {
+            let name = entry
+                .file_name()
+                .into_string()
+                .map_err(|bad| Error::NonUtf8SchemaDirEntry(bad.to_owned()))?;
+            let md =
+                entry.metadata().await.map_err(|err| Error::ReadSchemaDir {
+                    context: String::from("Failed to fetch entry metatdata"),
+                    err,
+                })?;
+            if !md.is_dir() {
+                debug!(log, "skipping non-directory"; "name" => &name);
+                continue;
+            }
+            match name.parse() {
+                Ok(ver) => {
+                    debug!(log, "valid version dir"; "ver" => ver);
+                    assert!(versions.insert(ver), "Versions should be unique");
+                }
+                Err(e) => warn!(
+                    log,
+                    "found directory with non-u64 name, skipping";
+                    "name" => name,
+                    "error" => ?e,
+                ),
+            }
+        }
+        Ok(versions)
+    }
+
+    /// Ensure that the database is upgraded to the desired version of the
+    /// schema.
+    ///
+    /// NOTE: This function is not safe for concurrent usage!
+    pub async fn ensure_schema(
+        &self,
+        replicated: bool,
+        desired_version: u64,
+        schema_dir: impl AsRef<Path>,
+    ) -> Result<(), Error> {
+        let schema_dir = schema_dir.as_ref();
+        let latest = self.read_latest_version().await?;
+        if latest == desired_version {
+            debug!(
+                self.log,
+                "database already at desired version";
+                "version" => latest,
+            );
+            return Ok(());
+        }
+        debug!(
+            self.log,
+            "starting upgrade to desired version {}", desired_version
+        );
+        let available = Self::read_available_schema_versions(
+            &self.log,
+            replicated,
+            schema_dir.clone(),
+        )
+        .await?;
+        // We explicitly ignore version 0, which implies the database doesn't
+        // exist at all.
+        if latest > 0 && !available.contains(&latest) {
+            return Err(Error::MissingSchemaVersion(latest));
+        }
+        if !available.contains(&desired_version) {
+            return Err(Error::MissingSchemaVersion(desired_version));
+        }
+
```

Comment:
Do we have a check making sure all available versions are consecutive numbers?

Reply:
No, we don't. What bad outcome would that protect us from, given that we only upgrade to versions strictly greater than the current?

Reply:
Perhaps this is a nit, but would like to avoid a situation where someone skipped a version by mistake (1, 2, 5...)

Reply:
Sure, done in fbd898397